theano/sandbox/mkl/mkl_type.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import absolute_import, print_function, division
+import os
+import numpy
+import theano
+from theano import Type, Variable
+from theano import tensor
+from six import StringIO
+
+try:
+    import mkl_ndarray
+    import mkl_ndarray.mkl_ndarray as mkl
+
+    try:
+        mkl_ndarray.__file__
+    except AttributeError:
+        from theano.gof.cmodule import get_lib_extension
+        mkl_ndarray.__file__ = os.path.join(mkl_ndarray.__path__._path[0], 'mkl_ndarray.' + get_lib_extension())
+except ImportError:
+    mkl = None
+
+
+class _operators(tensor.basic._tensor_py_operators):
+    dtype = property(lambda s: s.type.dtype)
+    broadcastable = property(lambda s: s.type.broadcastable)
+    ndim = property(lambda s: s.type.ndim)
+
+
+class MKLNdarrayVariable(_operators, Variable):
+    pass
+
+
+class MKLNdarrayType(Type):
```

Comment:
Before doing a full review of that, why do you want a new C object and not reuse numpy ndarray object or our libgpuarray object. libgpuarray object isn't just for GPU.

If there is a good reason to have a new C object defined in Theano, then I think we need to refactor that, as there is a big quantity of copied code from TensorType. You could make a GenericTensorType() object that both of them inherit from it, and maybe GpuArrayType (but we could do this in another PR).